Split package for poppler-qt

izahn opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->
closes https://github.com/conda-forge/poppler-feedstock/issues/113
<!--
Please add any other relevant info below:
-->
I've really struggled to get a split package working, appreciate any help!

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * The recipe must have some tests.

pkgw commented:
I think we might want to emulate the style of the `gobject-introspection` feedstock, which just has one build script where the resulting files are split. I'm playing around with trying that approach but haven't gotten it working just yet.

pkgw commented:
See #115. I was at least able to get it to build on linux-64, although it will probably need work for other OSs and cross builds.

izahn commented:
This is wonderful, thank you!

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipe**:

 * Jinja2 variable references are suggested to take a ``{{<one space><variable name><one space>}}`` form. See lines [162, 164].

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipe**:

 * Jinja2 variable references are suggested to take a ``{{<one space><variable name><one space>}}`` form. See lines [205, 207].

pkgw commented on recipe/meta.yaml:
```diff
@@ -64,16 +74,140 @@ requirements:
+      - {{ prefix }}share/man/man1/pdftops.1
+      - {{ prefix }}share/man/man1/pdftotext.1
+      - {{ prefix }}share/man/man1/pdfunite.1
+      - {{ prefix }}Library/bin/pdfattach.exe
```
Note that at the top of this file, `{{ prefix }}` is defined to include `Library/` on Windows. So you should be able to get rid of most of these; I think there is a variable `EXEEXT` or something that expands to `.exe` on Windows, for the files in `bin`?

izahn replied:
Oh I did miss that, thanks! I did see some error earlier that prompted me to add the `Library` prefix by hand, but I'll go back and clean it up.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

izahn commented:
I think this is starting to look pretty good except that the windows test fails. I can't understand what the error is trying to tell me. If you have time to take a look at the windows build log that would be awesome!

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * The "test" section was expected to be a dictionary, but got a CommentedSeq.

For **recipe**:

 * It looks like the 'poppler-qt' output doesn't have any tests.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

izahn commented:
Thanks to @pkgw this is ready for review!

xhochy commented:
From my learnings in the past, I would recommend to not use the files approach but rather use `conda-build`'s deduplication as we use in the `glib` feedstock. This is less brittle as we might otherwise omit files in the  `poppler` output that are added with newer versions. See https://github.com/conda-forge/glib-feedstock/blob/1bd2a53a047ca7cc52632b67ddf7d62fcc0e339f/recipe/install.sh#L13

I would add the following pseudo code to install.sh:

```
make install
if pkg_name == poppler; then
  remove poppler qt files
fi
```

That should be sufficient and get us rid of the explicit file listing for `poppler`.

izahn commented:
Thanks @xhochy that makes sense, I'll change to script based outputs.

One other doubt; should I set the version here to `21.03` since that is the current global pin in https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/eae23e911d84334ca1c233004b28841311fd091a/recipe/conda_build_config.yaml#L626 ? My goal is to be able to use `poppler-qt` to build other packages, e.g., in `staged-recipes` submissions.

izahn commented:
I switched to install scripts in https://github.com/conda-forge/poppler-feedstock/pull/114/commits/b1c121fcbeab86760df137ee16b9af5dd7ce7834 though I'm not entirely convinced that approach is better. It seems to me that either way we end up making  lists of files to include (or remove) and things will break it they change. I don't have a preference, feel free to keep or revert https://github.com/conda-forge/poppler-feedstock/pull/114/commits/b1c121fcbeab86760df137ee16b9af5dd7ce7834 as you prefer.

izahn commented:
Well, it turns out that my script-based approach failed on OSX. The problem seems to be that `ninja install` wants to re-run the build instead of just copying things as I expected it to. The build log is at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=378373&view=logs&jobId=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9. My preference would be to revert https://github.com/conda-forge/poppler-feedstock/commit/b1c121fcbeab86760df137ee16b9af5dd7ce7834 but I'm open to other suggestions.

izahn commented:
@xhochy I tried to get a script-based split package working, but it seemed even more fragile than the file-based method. I've reverted back to file-based here because that is the best I can currently get working.

After spending this time I'm personally not sure this is even worth it -- maybe easier to just have a totally separate `poppler-qt`, or just include it in a monolithic `poppler` package. I'm saying, split packages are kind of a pain...

pkgw commented:
@izahn Thanks for your persistence! This is close to working so I think it's worth pushing on through. I may try implementing @xhochy's suggestion and see if I can solve the problems you were having. One of these days we'll have a perfect package manager ...

izahn commented:
Thanks @pkgw and @xhochy appreciate all your help! Let me know if there is anything more I can do here.

izahn commented:
It looks good to me!

izahn commented:
Closing in favor of https://github.com/conda-forge/poppler-feedstock/pull/115
